x-pack/packages/kbn-elastic-assistant/impl/assistant/api/conversations/conversations.ts

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { OpenAiProviderType } from '@kbn/stack-connectors-plugin/common/openai/constants';
+import { HttpSetup } from '@kbn/core/public';
+import { IHttpFetchError } from '@kbn/core-http-browser';
+import {
+  ELASTIC_AI_ASSISTANT_CONVERSATIONS_URL,
+  ELASTIC_AI_ASSISTANT_API_CURRENT_VERSION,
+} from '@kbn/elastic-assistant-common';
+import { Conversation, Message } from '../../../assistant_context/types';
+
+export interface GetConversationByIdParams {
+  http: HttpSetup;
+  id: string;
+  signal?: AbortSignal | undefined;
+}
+
+/**
+ * API call for getting conversation by id.
+ *
+ * @param {Object} options - The options object.
+ * @param {HttpSetup} options.http - HttpSetup
+ * @param {string} options.id - Conversation id.
+ * @param {AbortSignal} [options.signal] - AbortSignal
+ *
+ * @returns {Promise<Conversation | IHttpFetchError>}
+ */
+export const getConversationById = async ({
+  http,
+  id,
+  signal,
+}: GetConversationByIdParams): Promise<Conversation | IHttpFetchError> => {
```

Comment:
Consider returning `Promise<Conversation>` instead of `Promise<Conversation | IHttpFetchError>` here and in similar functions in this file that return `Promise<Conversation | IHttpFetchError>`, because

- Returning a  `Promise<Conversation | IHttpFetchError>` requires the caller to imperatively branch on the return value after `await`ing the response
- When reading code that `await`s a function that returns `Promise<Conversation | IHttpFetchError>`, it's sometimes unclear whether or not the caller intended to ignore checking for errors, (when no branching logic is implemented) or if the developer was expecting uncaught errors to be handled in another part of the call stack

The following example from `x-pack/packages/kbn-elastic-assistant/impl/assistant/use_conversation/index.tsx` illustrates both of the above:

```ts
  /**
   * Updates the last message of conversation[] for a given conversationId with provided content
   */
  const amendMessage = useCallback(
    async ({ conversationId, content }: AmendMessageProps) => {
      const prevConversation = await getConversationById({ http, id: conversationId });
      if (isHttpFetchError(prevConversation)) {
        return;
      }
      if (prevConversation != null) {
        const { messages } = prevConversation;
        const message = messages[messages.length - 1];
        const updatedMessages = message ? [{ ...message, content }] : [];

        await appendConversationMessagesApi({
          http,
          conversationId,
          messages: updatedMessages,
        });
      }
    },
    [http]
  );
```

In the example code above, both calls:

```ts
const prevConversation = await getConversationById({ http, id: conversationId });
```

and

```ts
        await appendConversationMessagesApi({
          http,
          conversationId,
          messages: updatedMessages,
        });
```

return a

```ts
Promise<Conversation | IHttpFetchError>
```

and illustrate:

- Returning a  `Promise<Conversation | IHttpFetchError>` requires the caller to imperatively branch on the return value after `await`ing the response

This branching logic is illustrated by the following (from the example above):
```ts
      const prevConversation = await getConversationById({ http, id: conversationId });
      if (isHttpFetchError(prevConversation)) {
        return;
      }
      if (prevConversation != null)  {
      // ...
```

Both of the branches above appear to be unnecessary if the function returns a  `Promise<Conversation>`.

- When reading code that `await`s a function that returns `Promise<Conversation | IHttpFetchError>`, it's sometimes unclear whether or not the caller intended to ignore checking for errors, (when no branching logic is implemented) or if the developer was expecting unhandled error to be  caught / handled in another part of the call stack

This ambiguity is illustrated in the following branch:

```ts
      if (prevConversation != null) {
        const { messages } = prevConversation;
        const message = messages[messages.length - 1];
        const updatedMessages = message ? [{ ...message, content }] : [];

        await appendConversationMessagesApi({
          http,
          conversationId,
          messages: updatedMessages,
        });
      }
```

In the code above, it may not be obvious whether the author intended to ignore an `IHttpFetchError` returned by `appendConversationMessagesApi`, or if unhandled errors (in this function, as a result of calling `await appendConversationMessagesApi()`) were expected to propagate up the call stack. (They won't in this case, because an error is neither thrown, nor returned.)

Reply:
Resolved by commit https://github.com/elastic/kibana/pull/173487/commits/2ea2fb7a89d35ea7e3e920216ecb20b65210f1a3